internal/storage/streamed_chunk_uploader.go

```diff
@@ -0,0 +1,206 @@
+	Writing
+	// Upload failed.
+	WriteError
+	// Has been finalized.
+	Destroyed
+)
+
+// A StreamedChunkUploader is an implementation of ChunkUploader
+// interface, which use storage.Writer from go-storage-client
+// for resumable upload.
+//
+// It also stores state of how many bytes have been sent to it
+// to write, how many have been successfully written by it so far,
+// and the current state of the uploader.
+type StreamedChunkUploader struct {
+	// Implements the following interfaces.
+	gcs.ChunkUploader
+
+	// Internal objects for business logic
+	wc         *storage.Writer
+	objectName string
+
+	// Attributes for providing updates to user
+	totalWriteInitiatedSoFar int64
+	totalWriteSucceededSoFar int64
+	userProgressFunc         func(int64)
+
+	// Internal state for lifecycle management
+	state UploaderState
+}
+
+// NewStreamedChunkUploader creates a new instace of StreamedChunkUploader
+// for the given inputs.
+func NewStreamedChunkUploader(ctx context.Context, obj *storage.ObjectHandle, req *gcs.CreateObjectRequest, writeChunkSize int, progressFunc func(int64)) (*StreamedChunkUploader, error) {
+	if obj == nil || req == nil {
+		return nil, fmt.Errorf("nil obj or req")
+	}
+
+	if obj.ObjectName() != req.Name {
+		return nil, fmt.Errorf("names of passed object-handle (%s) and CreateObjectRequest (%s) don't match", obj.ObjectName(), req.Name)
+	}
+
+	if req.GenerationPrecondition != nil && *req.GenerationPrecondition != 0 {
+		return nil, fmt.Errorf("request received for pre-existing object %s, supported only for new objects", req.Name)
+	}
+
+	if writeChunkSize <= 0 {
+		return nil, fmt.Errorf("writeChunkSize <= 0 not supported")
+	}
+
+	// raw initialization
+	sow := StreamedChunkUploader{}
+	sow.state = Uninitialized
+
+	// store references to necessary parameters
+	sow.objectName = obj.ObjectName()
+	sow.userProgressFunc = progressFunc
+
+	// Create a NewWriter with the requested attributes, using Go Storage Client.
+	// Chuck size for resumable upload is default i.e. 16MB.
+	wc := obj.NewWriter(ctx)
+	// NewWriter never returns nil, so no nil-check here.
+
+	wc = storageutil.SetAttrsInWriter(wc, req)
+	wc.ChunkSize = writeChunkSize
+	wc.ProgressFunc = func(n int64) {
+		sow.totalWriteSucceededSoFar = n
+		logger.Debugf("%d bytes copied so far for object/file %s. chunk-size = %d", n, req.Name, wc.ChunkSize)
+
+		if sow.userProgressFunc != nil {
+			sow.userProgressFunc(n)
+		}
+	}
+	sow.wc = wc
+	sow.state = Initialized
+
+	return &sow, nil
+}
+
+func (sow StreamedChunkUploader) ChunkSize() (int, error) {
+	if nil == sow.wc {
+		return 0, fmt.Errorf("writer not initialized")
+	}
+	return sow.wc.ChunkSize, nil
+}
+
+func (sow StreamedChunkUploader) State() UploaderState {
+	return sow.state
+}
+
+func (sow StreamedChunkUploader) BytesUploadInitiatedSoFar() int64 {
+	return sow.totalWriteInitiatedSoFar
+}
+
+func (sow StreamedChunkUploader) BytesUploadedSoFar() int64 {
+	return sow.totalWriteSucceededSoFar
```

Comment:
why are these methods public, how will the caller use them?
I assume caller should use chunkUploader interface to upload and query the status.

Reply:
Are these methods only required in unit tests. If yes you should be able to access private methods since both the files are in same package. refer to bucket_handle_test

Reply:
`BytesUploadedSoFar()` is declared inthe ChunkUploader interface, so I've kept it public, and have made the non-interface functions private.